File: py_learn/fork.py
#测试fork子进程

import os

print('process id is:%s' % os.getpid())

pid = os.fork()
if pid == 0:
    print("i am child process, my pid is %s, my parent pid is %s" % (os.getpid(), os.getppid()))
else:
    print("i am father process:%s, my child pid is %s" % (os.getpid(), pid))

File: c_/fork.c
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <stdio.h>
#include <time.h>

void main()
{
	pid_t pid = fork();

	if (pid == 0) {
		printf("i am child, my ppid is %d\n", getppid());

		printf("STDERR_FILENO:%d, dtablesize:%d\n", STDERR_FILENO, getdtablesize());

		for (;;){
			sleep(1);
		}

	} else {
		printf("i am father, my child's pid is %d\n", pid);

		int st = 0;
		waitpid(pid, &st, 0);

		sleep(1);
		printf("child out ~st:%d\n", st);

		if (!WIFEXITED(st)) {
			printf("child exited abnormal~\n");
		} else {
			printf("exit status :%u\n", WEXITSTATUS(st));
		}
	}

}
